README.md
<img alt="NPM license"   src="https://img.shields.io/npm/l/jspanel4"> <img alt="npm version"   src="https://img.shields.io/npm/v/jspanel4?color=0677b8"> <img alt="npm downloads" src="https://img.shields.io/npm/dm/jspanel4?color=0677b8">


## [jsPanel 4.16.1 released 2022-11-03](#)

> As of v4.11.0-beta methods `jsPanel.ajax()` and `jsPanel.fetch()` are updated. That also affects options `contentAjax` and `contentFetch`. These updates might break existing code. So please check the docs on https://jspanel.de/

<img src="https://res.cloudinary.com/stefanstraesser-eu/image/upload/v1558601426/jsPanel4.7.0-sample-panels-1920_yzobd9.jpg">

**A dependency free javascript tool to create highly configurable multifunctional floating panels.**

+ a **huge amount of options** allows to configure appearance and behavior of a jsPanel
+ optionally add **extra toolbars** to header or/and footer sections
+ support for Boostrap (3 and 4) **themes** and Material-Design-for-Bootstrap themes is built-in
+ customize the panel controls with built-in support for **icon fonts** Font-Awesome, Material-Icons and Glyphicon
+ built-in support for **RTL text** direction
+ via extensions you can use it alose as **modal**, **tooltip**, **hint/notifier**, **contextmenu** or **dialog**
+ jsPanels are **draggable** and **resizable**, can be **maximized**, **minimized** and **smallified**
+ and much more ...

---

### jsPanel 4 homepage and documentation: [https://jspanel.de](https://jspanel.de/)

### [Dependencies]()
Just a modern mobile or desktop browser like FF, Chrome, EDGE, Brave, Opera, Vivaldi.
jsPanel 4 is pure javascript and does not depend on any other library.

### [Include the files]()
The following example shows a complete html file with the minimium setup:

```html
<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>jsPanel 4</title>
        <!-- loading jsPanel css -->
        <link rel="stylesheet" href="dist/jspanel.css">
    </head>
    <body>

        <!-- Your HTML goes here -->

        <!-- loading jsPanel javascript -->
        <script src="dist/jspanel.js"></script>
        <!-- optionally load jsPanel extensions -->
        <script src="dist/extensions/modal/jspanel.modal.js"></script>
        // and the other extension you need
    </body>
</html>
```

### [And then ...]()
After including all the necessary files in your project you can create a jsPanel like ...

```javascript
jsPanel.create( options );

// or
var myPanel = jsPanel.create( options );
```
... where **options** is an object passing the jsPanel configuration options to the function.

##### Example:

```javascript
jsPanel.create({
    position:    "left-top",
    contentSize: "600 350",
    contentAjax: {
    	url:  '../path/to/the/resource',
        done: function(xhr, panel) {
        	// the keyword "this" inside the function refers to the XMLHttpRequest object
        },
        fail: function(xhr, panel) {
        	//the keyword "this" inside the function refers to the XMLHttpRequest object
        }
    },
    headerTitle: "my example jsPanel",
    theme:       "rebeccapurple",
    callback:    function(panel) {
    	// do whatever you like
        // the keyword "this" inside the callback function refers to the panel
    }
});
```


CHANGELOG.md
## <span style='color:#563D7C;'>CHANGELOG</span>

### <span style='color:#563D7C;'>Version 4.16.1 *2022-11-03*</span>

+ simplified internal code to process theme options and removed some obsolete code
+ fix in css file to prevent overflowing title in minimized panel replacement

### <span style='color:#563D7C;'>Version 4.16.0 *2022-07-03*</span>

+ **added** Dialog extension. This extension adds an easy-to-use interface for dialog elements to virtually any jsPanel. It also offers a `modal()` function to create modal dialogs as well as `alert()`, `confirm()` and `prompt()` functions. `jsPanel.dialog` is a Third Party Extension developed and maintained by Michael Daumling.
+ **added** The object used to set a custom theme with `option.theme` now accepts the properties `bgFooter`, `colorFooter` and `borderRadius` as well
+ **added** method `jsPanel.getCssVariableValue()` to process css variables in options `theme`, `border`, `borderRadius` and methods `.setTheme()`, `.setBorder()`, `setBorderRadius()`
+ **updated** internal methods `.getThemeDetails()`, `.applyCustomTheme()` and `.pOborder()` to support the use of css variables in `option.theme`, `option.border`, `option.borderRadius` and the methods `.setTheme()`, `.setBorder()`, `setBorderRadius()`
+ **reorganized** code in the whole script

### <span style='color:#563D7C;'>Version 4.15.0 *2022-05-30*</span>

+ **added** option `css` to add custom css classes to the panel's html template, includes updated css files `jspanel.css`/`jspanel.min.css`
+ some internal code improvements

### <span style='color:#563D7C;'>Version 4.14.1 *2022-05-17*</span>

+ **fix** for maximized panels: on `resizestart` of a maximized panel its status is now reset to `'normalized'`, the _**normalize**_ control is hidden and the _**maximize**_ control is shown again.

### <span style='color:#563D7C;'>Version 4.14.0 *2022-04-20*</span>

+ **bugfix** in `close()` method (calling `close()` on an already closed panel caused an error)
+ **updated** option `onwindowresize`
+ **updated** option `onparentresize`

So far the two options `onwindowresize` and `onparentresize` could be set with either boolean `true` to start a preset behaviour or with a `function` to start a custom behaviour on the corresponding resize events. Now you can also set options `onwindowresize` and `onparentresize` with an object optionally allowing the use of both the preset behaviour AND an additional custom behaviour.

### <span style='color:#563D7C;'>Version 4.13.0 *2021-11-24*</span>

+ enabled default use of the PointerEvent API
+ **added** method `jsPanel.usePointerEvents()` to disable/enable use of PointerEvent API

### <span style='color:#563D7C;'>Version 4.12.0 *2021-07-09*</span>

+ updated CSS files to **bugfix** a minor issue in `option.minimizeTo`
+ **bugfix** in `resizeit` interaction 
+ **added** string value `'closed'` to panel property `status`. Since a closed panel is removed from the DOM this value is only available when a panel reference is saved.
+ **added** option to use a custom callback function as setting for `option.closeOnEscape` instead of just boolean true/false

### <span style='color:#563D7C;'>Version 4.11.4 *2021-04-10*</span>

+ **bugfix** in `dragit` interaction
+ **improved** panel positioning, dragit and snap features in order to allow for scrollbars
+ **added** option `parentPanelFront` to the tooltip configuration options (tooltip extension updated to v1.4.0)

### <span style='color:#563D7C;'>Version 4.11.3 *2021-02-03*</span>

+ **bugfix** in the code for the `resizeit` interaction reducing memory leaks. [See GitHub issue](https://github.com/Flyer53/jsPanel4/issues/121)
+ **bugfix** when using the **dock** extension a _docked_ panel:
  + did not reposition when the _master_ panel was repositioned programmatically
  + did not reposition when the _master_ panel was resized programmatically
  + did not reposition when **option** `onwindowresize` is activated and the window is resized
+ **bugfix** `layout` extension: `jsPanel.layout.restoreId` now returns the restored panel
+ **added** parameter `closeOnMouseleave` in `contextmenu` extension

### <span style='color:#563D7C;'>Version 4.11.2 *2020-12-09*</span>

+ **bugfix** in the code for the `dragit` interaction reducing memory leaks. [See GitHub issue](https://github.com/Flyer53/jsPanel4/issues/121)

### <span style='color:#563D7C;'>Version 4.11.1 *2020-11-23*</span>

+ **bugfix** in the `maximize()` method

### <span style='color:#563D7C;'>Version 4.11.0 *2020-11-20*</span>

+ **renamed** panel property `autocloseProgressbar` to `progressbar`. The old property name remains usable for compatibility.
+ **bugfix** in option `onwindowresize`
+ various internal improvements in js and css

**See also the changelog of _v4.11.0-beta_ for a list of changes compared with _v4.10.2_**

### <span style='color:#563D7C;'>Version 4.11.0-beta *2020-09-06*</span>

+ **added** method `jsPanel.strToHtml(str)` converting a DOMString to a `DocumentFragment`.
+ **added** option `opacity`
+ **bugfix** in _modal_ extension concerning the use of `onclosed` callback.
+ **bugfix** in `autoposition` parameter of option `position`.
+ **bugfix** in option `onwindowresize`.
+ **updated** `jsPanel.ajax()` in order to improve its usability as general purpose AJAX tool.<br>**This update might require a change in your code if you use `jsPanel.ajax()` and/or option `contentAjax`.**
+ **updated** `jsPanel.fetch()` in order to improve its usability as general purpose Fetch tool.<br>**This update might require a change in your code if you use `jsPanel.fetch()` and/or option `contentFetch`.**
+ **updated** option `contentAjax` optionally loads only a page fragment.
+ **updated** CSS for controlbar
+ **updated** header title is now wrapped in a `<div>` instead of a `<span>`
+ **updated**  _contextmenu_ extension due to changes in `jsPanel.ajax()`.
+ **updated** dock extension internally. Should not affect existing code.

source/extensions/dialog/jspanel.dialog.js
if (!jsPanel.dialog) {
	jsPanel.dialog = {
		version: '1.0.0',
		date: '2022-04-25 10:00',
		defaults: {
			theme: "none", 
			header: false,
			position: {my:'center-top', at:'center-top', offsetY: 30},
			contentSize: "auto",
			onwindowresize: true,
			closeOnEscape: false,
			closeOnBackdrop: false,
			oninitialize: []
		},
		// You can wrap dialogs into this <template> tag to prevent them from rendering
		dialogTemplateId: "#dialogs",
		// CSS classes
		css: {
			// primary button
			primaryBtn: "blue",
			// all other buttons
			otherBtn: "white",
			// the button bar
			buttonBar: "buttonbar",
			// the input field in the prompt() alert
			promptInput: "prompt-input"
		},
		// The vertical offset for child dialogs
		offsetY: 30,

		/**
		 * Display a modal dialog. The html may either be a selector (whose content
		 * is deep-cloned), a HTML string, a HTMLElement (deep-cloned), or a simple
		 * string, which is wrapped in a <span> tag. Since existing HTML is cloned,
		 * you should not use IDs in that HTML, because these IDs will occur twice
		 * as long as the dialog exists.
		 * 
		 * You can also wrap your dialogs in a <template> tag, which keeps them from
		 * being rendered. The ID of that tag is stored in jsPanel.dialog.templateId
		 * (the default is "dialogs"). When you do this, you are also free to use
		 * HTML IDs if required.
		 * 
		 * To initialize the dialog, use either the "callback" option, which is called
		 * synchronously during the creation of the panel, or use the "oninitialize" 
		 * option, which takes a function or an array of functions as usual. This 
		 * callback is called after the panel has been created, so it is safe to e.g. 
		 * display alerts inside the function. 
		 * 
		 * The modal's value is the dialog value (see makeDialog() below).
		 * 
		 * @param {string|HTMLElement} html (a selector, a HTML string, or the HTML itself)
		 * @param {Object} options 
		 * @returns {*} any value returned by a handler, or the value of the last clicked element
		 */
		async modal(html, options = {}) {
			html = this.helpers.getHTML(html);
			options = Object.assign({}, this.defaults, options);
			options.content = html;
			const prev = this.active;
			if (prev) {
				options.position = Object.assign({}, options.position);
				options.position.offsetY = prev.options.position.offsetY + this.offsetY;
			}
			options.css = { panel: "jsPanel-dialog" };
			for (let cls of ["dialog-sm", "dialog-md", "dialog-lg", "dialog-xl"]) {
				if (!html.classList.contains(cls))
					continue;
				html.classList.remove(cls);
				options.css.panel += " " + cls;
				break;
			}
			// Make sure that the callback function is an array
			if (!options.callback)
				options.callback = [];
			else if (typeof options.callback === "function")
				options.callback = [options.callback];
			options.callback.unshift(panel => {
				panel.makeDialog();
				if (this.helpers.all.length)
					// Move a previous modal dialog into the background
					this.helpers.all[0].classList.add("background");
				this.helpers.all.unshift(panel);
			});
			document.addEventListener("jspanelloaded", ev => {
				if (ev.panel.options.oninitialize)
					jsPanel.processCallbacks(ev.panel, ev.panel.options.oninitialize, "every");
			}, { once: true });
			const panel = jsPanel.modal.create(options);
			panel.resize({height: 'auto'});
			return new Promise(resolve => {
				panel.options.onclosed.push(panel => {
					this.helpers.all.shift();
					if (this.helpers.all.length)
						this.helpers.all[0].classList.remove("background");
					const value = panel.dialog.value;
					panel.dialog.value = undefined;
					resolve(value);
				});
			});
		},
		/**
		 * Return the currently active modal dialog.
		 */
		get active() {
			return this.helpers.all[0];
		},

		/**
		 * Return the number of open dialogs.
		 */
		get depth() {
			return this.helpers.all.length;
		},

		/**
		 * Force-close all active dialogs.
		 */
		closeAll() {
			while (this.helpers.all.length) {
				const len = this.helpers.all.length;
				this.helpers.all[0].close();
				if (this.helpers.all.length === len)
					// close refused
					return;
			}
			// sometimes, there seems to bea leftover backdrop
			for (let el of document.querySelectorAll(".jsPanel-modal.background"))
				el.remove();
		},

		/**
		 * Display a message box with the given list of buttons.
		 * If there is no button, create an OK button. The buttons array
		 * is either a list of labels, or objects with these properties:
		 * label - the label
		 * value - the value for panel.dialog.value when clicked (opt)
		 * name - the name of the button if they are accessed ba name (opt)
		 * css - any CSS classes for the button
		 * 
		 * @param {string|Node} msg (can be a string, a HTML string, or HTML)
		 * @param {Array} buttons 
		 * @param {Object} options
		 * @returns {string} the button value
		 */
		async alert(msg, buttons = ["OK"], options = {}) {
			msg = this.helpers.getHTML(msg);
			if (!buttons.length)
				buttons.push("OK");
			msg.append(this.helpers.buttonBar(buttons));
			return this.modal(msg, options);
		},

		/**
		 * Display a Confirm (Yes/No) box. If no is true, No is set as the default.
		 * 
		 * @param {string} msg the message
		 * @param {bool} no if true, No is preselected
		 * @param {Array} moreButtons any extra buttons (see jsPanel.dialog.alert())
		 * @param {Object} options
		 * @returns {bool}
		 */
		async confirm(msg, no = false, moreButtons = [], options = {}) {
			const y = { label: "Yes", value: "true" };
			const n = { label: "No", value: "false" };
			let btns = no ? [n, y] : [y, n];
			btns = btns.concat(moreButtons);
			return await this.alert(msg, btns, options) === "true";
		},

		/**
		 * Display a prompt box with a message and an input field, with an optional
		 * preset. Either returns the content of the input field or null.
		 * @param {String} msg
		 * @param {String} preset
		 * @param {Object} options
		 * @returns {string|null}
		 */
		async prompt(msg, preset = "", options = {}) {
			msg = this.helpers.getHTML(`<div>${msg}</div>`);
			const div = document.createElement("div");
			if (msg instanceof DocumentFragment) {
				for (let el = msg.firstChild; el; el = el.nextSibling)
					div.append(msg);
			}
			else
				div.append(msg);
			const input = jsPanel.strToHtml(`<input name="input" type="text" class="${this.css.promptInput}" value="${preset}"/>`);
			div.append(input.firstElementChild);
			div.append(this.helpers.buttonBar(["OK", "Cancel"]));
			options = Object.assign({
				onclick_OK: panel => panel.dialog.value = panel.dialog.values.input,
				onclick_Cancel: async panel => panel.dialog.value = null
			}, options);
			return await this.modal(div, options);
		},

		helpers: {
			// All open modal dialogs
			all: [],
			/**
			 * Create HTML for the given buttons array. Each element is an object with 
			 * label and value. The buttons are dismiss buttons. The only button (or 
			 * the second button) is marked as the Cancel button that acts on the Esc 
			 * key. We also have the "css" property that, optionally, defines button 
			 * css classes, and a "name" property if the button should be named.
			 * 
			 * If the array element is just a string, name and value are set to that string.
			 * 
			 * @param {Array} buttons 
			 * @returns {HTMLElement}
			 */
			buttonBar(buttons) {
				const div = document.createElement("div");
				div.classList.add(jsPanel.dialog.css.buttonBar);
				let html = "";
				for (let [i, btn] of buttons.map(b => (typeof b === "string") ? { label:b,name:b,value:b} : b).entries()) {
					let {label, value, css, name} = btn;
					const classDefs = jsPanel.dialog.css;
					if (!css)
						css = i ? classDefs.otherBtn : classDefs.primaryBtn;
					let cancel = "";
					name = name ? `name="${name}"` : "";
					switch (i) {
						case 0: 
							if (buttons.length === 1)
								cancel = " data-cancel";
							break;
						case 1:
							cancel = " data-cancel";
							break;
					}
					html += `<button data-dismiss ${name} ${cancel} class="${css}" value="${value}">${label}</button>`;
				}
				div.innerHTML = html;
				return div;
			},

			/**
			 * Get the HTML to display - see modal() above and make sure it is visible.
			 * @param {Node|string} html 
			 * @returns {Node}
			 */
			getHTML(html) {
				if (!(html instanceof Node)) {
					html = html.toString().trim();
					try {
						let el = document.querySelector(html);
						if (!el) {
							const tpl = document.querySelector(jsPanel.dialog.dialogTemplateId);
							if (tpl)
								el = tpl.content.querySelector(html);
						}
						if (el)
							html = el.cloneNode(true);
						else
							html = jsPanel.strToHtml("<span>" + html + "</span>");
					}
					catch (e) {
						html = jsPanel.strToHtml(html.toString().trim());
					}
				}
				if (html instanceof DocumentFragment) {
					// convert a DocumentFragment to a div, because the former cannot have CSS
					let div = document.createElement("div");
					for (let node of html.childNodes)
						div.append(node);
					return div;
				}
				if (html instanceof HTMLElement)
					// ensure its visibility
					html.style.display = "";
				return html;	
			},

			/**
			 * Get the value of an element.
			 * 
			 * The values are:
			 * radio buttons: the value of the selected button
			 * checkboxes: true or false
			 * text etc: the text
			 * elements with a "value" property: the value of that property
			 * everything else: the inner HTML
			 * 
			 * @param {jsPanel} panel
			 * @param {HTMLElement} element
			 * @param {bool} returnHTML - false to return undefined for no value
			 */
			getValue(panel, el, returnHTML = true) {
				switch (el.getAttribute("type"))
				{
					case "radio":
						const name = el.getAttribute("name");
						const radio = panel.querySelector(`[name="${name}"]:checked`);
						return radio ? radio.value : "";
					case "checkbox":
						return el.checked;
					default:
						if ("value" in el)
							return el.value;
						else if (el.hasAttribute("value"))
							return el.getAttribute("value");
						else if (returnHTML)
							return el.innerHTML;
				}
			},
			/**
			 * Set the value of an element.
			 * 
			 * The values are:
			 * radio buttons: the value of the selected button
			 * checkboxes: true or false
			 * text etc: the text
			 * elements with a "value" property: the value of that property
			 * everything else: the inner HTML (lets you set content)
			 * 
			 * @param {jsPanel} panel
			 * @param {HTMLElement} element
			 * @param {*} value
			 */
			setValue(panel, el, value) {
				switch(el.getAttribute("type")) {
					case "radio":
						const name = el.getAttribute("name");
						const radio = panel.querySelector(`[name="${name}"][value="${value}"]`);
						if (radio)
							radio.checked = true; 
						break;
					case "checkbox": 
						el.checked = value; 
						break;
					default:
						if ("value" in el)
							el.value = value;
						else if (el.hasAttribute("value"))
							el.setAttribute("value", value.toString());
						else
							el.innerHTML = value.toString();
				}
			}
		}
	};

	document.addEventListener("click", ev => {
		// handle all clicks on elements without a "name" property;
		// elements with a "name" property are handled in jsPanel.makeDialog()
		let el = ev.target.closest("[data-dismiss]");
		if (!el)
			return;
		let panel = el.closest(".jsPanel");
		if (!panel)
			return;
		if (el.click)
			el.click();
		if (panel.parentElement)
			panel.close();
	});

/*	window.addEventListener("resize", ev => {
		for (let panel of jsPanel.dialog.helpers.all)
			panel.reposition();
	});
*/
	jsPanel.extend({
		/**
		 * Extend any jsPanel to act as a dialog.
		 * 
		 * Elements of this panel can have several additional attributes:
		 * 
		 * The "name" attribute is well-known from forms. Actually, you can set this
		 * attribute at any element. Elements with a "name" attribute have
		 * several advantages:
		 * 
		 * The panel property "dialog.elements" refers any elements with a "name"
		 * attribute. For example, if you have a textfield with name="user", you
		 * can access that element as panel.dialog.elements.user. If multiple elements
		 * share the same value, the value of this property is a NodeList.
		 * 
		 * The panel property "dialog.values" permits direct access to an element's
		 * value. In the above example, panel.dialog.values.user gets and sets the
		 * text of that element. If the element does not have a direct value, you
		 * can define a value with the "value" attribute. If the element does not
		 * have a "value" attribute, you can get and set the element's inner HTML
		 * this way.
		 * 
		 * The "data-dismiss" attribute causes the panel to be closed when that element
		 * is clicked.
		 * 
		 * The "data-cancel" attribute causes the element to respond to the "Escape"
		 * key. Note that the closeOnEscape option is not used; instead, an own event
		 * listener handles the Escape key, because the closeOnEscape option closes 
		 * just any panel that happens to return true in the handler, which is not 
		 * desirable in a dialog environment.
		 * 
		 * The "data-dblclick" causes the element to respond to a double click as
		 * if it was clicked.
		 * 
		 * The panel options can define several handlers for elements. Each handler
		 * starts with either "onclick_" for clicks, or "oninput_" for changes to
		 * a text element, followed by the element's name, like e.g. "oninput_user".
		 * It is fine to call panel.close() in a handler.
		 * 
		 * All handlers can set a value by setting panel.dialog.value. If a modal
		 * dialog was created, the return value of the await statement is that
		 * value. If no handler was defined for a clicked element, the element's "value" 
		 * attribute is set as the dialog value. Inside a handler, you can, of course,
		 * read panel.dialog.value.
		 */
		makeDialog() {
			this.dialog = {
				elements: new jsPanelDialogElements(this),
				values: new jsPanelDialogValues(this),
				value: undefined
			};
			const cb = async ev => {
				let el = ev.target.closest("[name]");
				ev.stopPropagation();
				if (el) {
					const name = el ? el.getAttribute("name") : "";
					const cbName = `on${ev.type}_${name}`;
					if (this.options[cbName])
						jsPanel.processCallbacks(this, this.options[cbName], "every", el, ev);
				}
				else
					el = ev.target;
				if (el.hasAttribute("data-dismiss") && this.parentElement) {
					// grab any value if not set yet
					if (this.dialog.value === undefined)
						this.dialog.value = jsPanel.dialog.helpers.getValue(this, el, false);
					this.close();
				}
			};
			this.addEventListener("click", cb);
			this.addEventListener("input", cb);
			this.addEventListener("dblclick", ev => {
				const el = this.querySelector("[data-dblclick]");
				if (el) {
					ev.stopPropagation();
					el.click();
				}
			});
			this.options.closeOnEscape = _ => {
				// Trigger a data-cancel element if present
				const el = this.querySelector("[data-cancel]");
				if (el)
					el.click();
				return false;
			}
		}
	});

	/**
	 * This class acts as a proxy for all dialog elements with either
	 * a "name" attribute or a "data-name" attribute. You can address
	 * such an element with panel.dialog.values.xxx, where xxx is the
	 * element's name.
	 * 
	 * The object is enumerable; you may walk over the object with the
	 * usual helpers. If multiple elements have the same name, the getter
	 * returns an array; otherwiese, it returns the HTMLElement or null.
	 */
	class jsPanelDialogElements {
		constructor(panel) {
			for (let el of panel.querySelectorAll("[name],[data-name]")) {
				const name = el.getAttribute("name") || el.getAttribute("data-name");
				if (this.hasOwnProperty(name))
					continue;
				Object.defineProperty(this, name, {
					enumerable: true,
					get() {
						const result = panel.querySelectorAll(`[name="${name}"],[data-name="${name}"]`);
						return result.length <= 1 ? result[0] : result;
					}
				});
			}
			Object.seal(this);
		}
		/**
		 * Retrieve an object with key/value pairs for all elements. This
		 * is more versatile than using a FormData object, because it can
		 * also contain numbers, booleans etc.
		 */
		get() {
			const obj = {};
			for (let [name, value] of Object.entries(this))
				obj[name] = value;
			return obj;
		}
	}

	/**
	 * This class acts as a proxy for all dialog elements with 
	 * a "name" attribute. You can address such an element's value with 
	 * panel.dialog.values.xxx, where xxx is the element's name. 
	 * 
	 * The values are:
	 * radio buttons: the value of the selected button
	 * checkboxes: true or false
	 * text etc: the text
	 * elements with a "value" property: the value of that property
	 * everything else: the inner HTML (lets you set content)
	 * 
	 * The object is enumerable; you may walk over the object with the
	 * usual helpers. 
	 * 
	 * If multiple elements have the same name, only the first element
	 * is affected, except for radio buttons.
	 */
	class jsPanelDialogValues {
		constructor(panel) {
			for (let el of panel.querySelectorAll("[name]")) {
				const name = el.getAttribute("name");
				if (this.hasOwnProperty(name))
					continue;
				const type = (el.getAttribute("type") || "").toLowerCase();
				Object.defineProperty(this, name, {
					enumerable: true,
					get() { return jsPanel.dialog.helpers.getValue(panel, el); },
					set(value) { jsPanel.dialog.helpers.setValue(panel, el, value); }
				});
			}
			Object.seal(this);
		}

		/**
		 * Retrieve an object with key/value pairs for all elements. This
		 * is more versatile than using a FormData object, because it can
		 * also contain numbers, booleans etc.
		 */
		get() {
			const obj = {};
			for (let [name, value] of Object.entries(this))
				obj[name] = value;
			return obj;
		}
		/**
		 * Use an object with key/value pairs to set element values.
		 * 
		 * @param {Object} obj
		 */
		set(obj) {
			for (let [name, value] of Object.entries(obj))
				this[name] = value;
		}
	}
}
